### DISCUSSION TITLE
Guide to compiling 'Hello World' using BlitzMax and VS Code

### DISCUSSION THREAD

**Discussion by Hezkore** [Hezkore/vscode-blitzmax-support#10] [Guides]
### Tl;dr version
Install [BlitzMax NG](https://blitzmax.org/downloads/) and [VS Code](https://code.visualstudio.com/download)
Open or create a **.bmx** file in VS Code _(Open file / Open folder are both okay)_
<kbd>Ctrl</kbd> + <kbd>P</kbd> > **ext install Hezkore.blitzmax**
Press one of the quick build buttons in the corner, or <kbd>F5</kbd> for a quick debug build run

### Step by step
1. Download BlitzMax NG for your platform and _remember where you install it!_ - https://blitzmax.org/downloads
2. Setup BlitzMax NG for your platform - [Windows](https://blitzmax.org/docs/en/setup/win32/), [Linux](https://blitzmax.org/docs/en/setup/linux/), [MacOS](https://blitzmax.org/docs/en/setup/macos/), [Raspberry Pi](https://blitzmax.org/docs/en/setup/rpi/)
3. Download VS Code for your platform - https://code.visualstudio.com/download
4. Start VS Code and click on **File** > **New File**
5. Type `Print "Hello World"` in your new empty file
6. Click on **File** > **Save** and save your file as **hello_world.bmx** _(it's important the file ends with **.bmx**)_
7. Click on **View** > **Extensions** and search for **BlitzMax** and click **Install**
8. You'll be notified that **"No BlitzMax path configured"**, click **Select path** and select the where you previously installed BlitzMax
_(i.e. `c:\blitzmax\`)_
9. Jump back to your **hello_world.bmx** file by pressing the tab at the top of the VS Code window
10. Press <kbd>F5</kbd> to debug compile your **hello_world.bmx** file and run it
11. You will now see `Hello World` at the bottom of the screen in the **Debug Console**

#### What now?
Learn how to use BlitzMax NG - https://blitzmax.org/docs/en/tutorials/beginners_guide
Learn how to use VS Code - https://code.visualstudio.com/docs/getstarted/introvideos

Remember that the initial compile time might be long, as the BlitzMax modules hasn't been compiled yet.
You can cut down on compile time by setting a framework
```BlitzMax
Framework brl.standardio
Print "Hello World"
```

When opening **.bmx** files, you might be notified that VS Code has an extension for **.bmx** files
It's okay to follow this notification, it will install the same extension

Use the **BlitzMax** view to the left _(rocket icon)_ to configure your build options
Double click any option in the list to toggle it on or off
The **Legacy Mode** build option enables build support for BlitzMax Vanilla/Legacy
Hover over the other options with your mouse for a quick explanation

Remember that pressing <kbd>F5</kbd> is the VS Code shortcut to **"Start Debugging"**, so this will always result in a debug console build
For a release build, set the proper build options and use the quick build buttons in the upper right corner when viewing any **.bmx** file
You can also use VS Code **tasks** to make your life a whole lot easier, and set up some very complex build chains
When you eventually feel ready for tasks, you can read the extensive task guide https://github.com/Hezkore/vscode-blitzmax-support/discussions/12

You might get a notification saying **"Documentation not found"**
This happens when the BlitzMax modules hasn't been documented yet 
You should press **Yes** when promoted to rebuild them

When documentation has been built, the following features will work correctly:
* Hover your mouse over any known keyword or function to get useful information
* Click on any known keyword or function, then press <kbd>F1</kbd> for full documentation and examples
* Press <kbd>Ctrl</kbd> + <kbd>Space</kbd> to trigger auto complete
* Press <kbd>Tab</kbd> in the auto complete list to quickly insert a function or keyword, with function parameter template\
_(press <kbd>Tab</kbd> again to jump to the next function parameter)_
* Press <kbd>Ctrl</kbd> + <kbd>Shift </kbd> + <kbd>Space</kbd> inside a function parameter to get symbol help

### Existing BlitzMax projects
For existing BlitzMax projects, all you have to do is select **File** > **Open Folder** and select the root folder of your project
Click on the **Explorer** view to the left _(or **View** > **Explorer**)_ to easily navigate your project files
This works well with the examples projects that comes with BlitzMax, as well as your own projects
To select/lock the file to build, you can right click any **.bmx** file and select **Set As Default Build Task Source File**
_(this option is only available for **Open Folder** projects or workspaces)_
